### PR TITLE
Ensure latest canary of Next.js is installed in repros

### PR DESCRIPTION
**Pull request by karlhorky**
### What?

Always use latest canary in reproduction templates

### Why?

Following up with #63038, it appears that it has not completely resolved the issue with CodeSandbox caching lockfiles and using an outdated version of Next.js

Eg. currently, the reproduction template URL uses `next@14.2.0-canary.9` is not the latest canary

https://codesandbox.io/p/sandbox/github/vercel/next.js/tree/canary/examples/reproduction-template

### How?

This PR implements the following changes to the reproduction template to install the latest version on fork:

1. On CodeSandbox "dev" task start, always update to latest Next.js with `pnpm update next`
2. Make the CodeSandbox "dev" task restart on fork / clone of the sandbox (@CompuIves  built this feature very recently)

Observe how the `pnpm update next` script at the start of the `dev` task runs again when forking my branch of the reproduction example:

https://github.com/vercel/next.js/assets/1935696/a3f6e65c-02cb-4eac-b0fa-367ac1751c39

cc @samcx @styfle @CompuIves

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 2c4872dae57da7c31cfcedae2ec2d323e13d36c3

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by karlhorky** [comment]
@samcx the `restartOn.clone` feature that this PR relies upon has been published to CodeSandbox now, so this should be ready for review.

I added a video of the behavior of forking, the "Already up to date" message at the top of the terminal output is the result of running `pnpm update next`. We won't be able to see the CodeSandbox forking behavior actually installing a new version until a new canary is released or this PR is merged.

File: examples/reproduction-template-pages/.codesandbox/tasks.json
Comment (styfle) [inline examples/reproduction-template-pages/.codesandbox/tasks.json:11]: The PR is titled "Ensure latest canary" but wont this update to stable only? It would need to be `next@canary` instead.

Comment (karlhorky) [reply examples/reproduction-template-pages/.codesandbox/tasks.json:11]: Given the `"next": "canary"` dependency in `package.json` in the reproduction templates, `pnpm update next` updates next to the latest canary at that time:

https://github.com/vercel/next.js/assets/1935696/2bbcfbfc-bee0-4cd7-ab15-dca3b70a44fe

Comment (karlhorky) [reply examples/reproduction-template-pages/.codesandbox/tasks.json:11]: This means that the version number will be tied to the forked reproduction sandbox, which is often wanted, and avoids the following issue:

1.⁠ User creates a reproduction on a certain canary version
2.⁠ ⁠⁠New canary version is released
3.⁠ ⁠⁠User restarts the `dev` task for any reason
4.⁠ ⁠⁠Next.js version updated, potentially leading to the bug going away 💥

**Review by styfle — APPROVED** [review]

**Comment by karlhorky** [comment]
@styfle Thanks for the review and merge!

I can confirm that both of the [App Router reproduction template](https://codesandbox.io/p/sandbox/github/vercel/next.js/tree/canary/examples/reproduction-template) and the [Pages Router reproduction template](https://codesandbox.io/p/sandbox/github/vercel/next.js/tree/canary/examples/reproduction-template-pages) now have the desired behavior:

1. On fork of either reproduction template, `pnpm update next` is run, updating to the latest canary version (pinned) in the new forked sandbox

**Comment by karlhorky** [comment]
Oh, looks like I still haven't been able to avoid the CodeSandbox caching behavior:

1. When visiting https://codesandbox.io/p/sandbox/github/vercel/next.js/tree/canary/examples/reproduction-template?file=%2Fpackage.json, CodeSandbox redirects to [a devbox called `blazing-wood-go8s7s` without an owner](https://codesandbox.io/p/devbox/blazing-wood-go8s7s?file=%2Fpackage.json) which includes a cached `package.json` with the pinned `14.3.0-canary.26` version of `next`
2. The cached `package.json` with the pinned Next.js version is also visible in the alternative `/devbox/` link https://codesandbox.io/p/devbox/github/vercel/next.js/tree/canary/examples/reproduction-template?file=%2Fpackage.json

This cached devbox means the reproduction will still get a pinned, cached version of Next.js

Ives has confirmed problems with the behavior above - we'll investigate minimal ways to disable this caching behavior. (Ideally completely on the CodeSandbox side, but potentially also involving the config here)

Eg. first thoughts: Maybe some environment variable that I can use to avoid creation of this extra devbox, which caches the `package.json` with the pinned version, eg:

```bash
# Only run `pnpm update next` when the user has owner rights on the sandbox
([[ -z “${CODESANDBOX_HAS_OWNER_RIGHTS}” ]] && pnpm update next) && pnpm dev
```

**Comment by karlhorky** [comment]
Ok, haven't come up with a good final solution yet that avoids [the papercut / unexpected behavior of restarting the `dev` task](https://github.com/vercel/next.js/pull/64967#discussion_r1581120614), but instead of waiting longer for us to come up with a more elegant solution for CodeSandbox, opening a PR to switch to `pnpm update next@canary` (previously [suggested above](https://github.com/vercel/next.js/pull/64967#discussion_r1581108225)), which upgrades the pinned version in `package.json`, and will restart on forking of the sandbox (because of `restartOn.clone = true`):

```bash
$ pnpm update next@canary && cat package.json
Packages: +4 -4
++++----
Progress: resolved 65, reused 43, downloaded 0, added 0, done

dependencies:
- next 14.3.0-canary.26
+ next 14.3.0-canary.32

Done in 2.9s
```

Opened a minimal PR with this change:

- https://github.com/vercel/next.js/pull/65197

File: examples/reproduction-template-pages/.codesandbox/tasks.json
Comment (karlhorky) [reply examples/reproduction-template-pages/.codesandbox/tasks.json:11]: Hm, looks like currently, [the CodeSandbox caching behavior (saving the results of `pnpm update next` in the `package.json` in the devbox)](https://github.com/vercel/next.js/pull/64967#issuecomment-2083480368) is leading to behavior which makes me lean more towards `pnpm update next@canary` again:

- https://github.com/vercel/next.js/pull/64967#issuecomment-2084963925

**Pull request merged**
